src/lib/SaferNames/Name.hs

```diff
@@ -0,0 +1,478 @@
+-- `traverseNamesB` and similar.
+data FreshBinder (b::B) (n::S) (i::S) where
+  FreshBinder :: FreshExt o o' -> b o o' -> RenameEnv i o'
+              -> FreshBinder b o i
+
+class HasNamesB (b::B) where
+  traverseNamesB :: Monad m => Scope o -> RenameTraversal m i o
+                       -> b i i'
+                       -> m (FreshBinder b o (i:=>:i'))
+  boundScope :: b n l -> Scope (n:=>:l)
+
+-- slightly safer than raw `unsafeCoerce` because at least it checks the kind
+unsafeCoerceE :: forall (e::E) i o . e i -> e o
+unsafeCoerceE = unsafeCoerce
+
+unsafeCoerceB :: forall (b::B) n l n' l' . b n l -> b n' l'
+unsafeCoerceB = unsafeCoerce
+
+withFresh :: Scope n -> (forall l. FreshExt n l -> PlainBinder n l -> Name l -> a) -> a
+withFresh (UnsafeMakeEnv scope) cont =
+  cont UnsafeMakeExt (UnsafeMakeBinder freshName) (UnsafeMakeName freshName)
+  where freshName = freshRawName "v" (LM.keysSet scope)
+
+freshRawName :: D.Tag -> S.Set RawName -> RawName
+freshRawName tag usedNames = D.Name D.GenName tag nextNum
+  where
+    nextNum = case S.lookupLT (D.Name D.GenName tag bigInt) usedNames of
+                Just (D.Name D.GenName tag' i)
+                  | tag' /= tag -> 0
+                  | i < bigInt  -> i + 1
+                  | otherwise   -> error "Ran out of numbers!"
+                _ -> 0
+    bigInt = (10::Int) ^ (9::Int)  -- TODO: consider a real sentinel value
+
+unitName :: Name UnitScope
+unitName = UnsafeMakeName "TheName"
+
+unitScope :: Scope UnitScope
+unitScope = UnsafeMakeEnv $ LM.singleton unitRawName ()
+  where UnsafeMakeName unitRawName = unitName
+
+envMapWithKey :: (Name i -> a -> b) -> Env i a -> Env i b
+envMapWithKey f (UnsafeMakeEnv m) = UnsafeMakeEnv $ LM.mapWithKey f' m
+  where f' rawName = f $ UnsafeMakeName rawName
+
+envPairs :: Env (n:=>:l) a -> (PlainBinderList n l, [a])
+envPairs (UnsafeMakeEnv m) = do
+  let (names, vals) = unzip $ LM.assocs m
+  (unsafeMakePlainBinderList names, vals)
+
+unsafeMakePlainBinderList :: [RawName] -> PlainBinderList n l
+unsafeMakePlainBinderList names = case names of
+  [] -> unsafeCoerceB Empty
+  (name:rest) -> Nest (UnsafeMakeBinder name) $ unsafeMakePlainBinderList rest
+
+-- === specialized traversals ===
+
+type NameSet n = S.Set (Name n)
+
+fmapNames :: HasNamesE e => Scope b -> (Name a -> Name b) -> e a -> e b
+fmapNames scope f e = runIdentity $ traverseNamesE scope t e
+  where t = newNameTraversal (pure . f)
+
+foldMapNames :: (HasNamesE e, Monoid a) => (Name n -> a) -> e n -> a
+foldMapNames f e =
+  execWriter $ traverseNamesE unitScope t e
+  where t = newNameTraversal \name -> tell (f name) >> return unitName
+
+freeNames :: HasNamesE e => e n -> NameSet n
+freeNames e = foldMapNames S.singleton e
+
+-- should be equivalent to `\ext -> fmapNames (injectName ext)`
+injectNamesL :: HasNamesE e => FreshExt n l -> e n -> e l
+injectNamesL _ = unsafeCoerceE
+
+projectNamesL :: HasNamesE e => Scope n -> Scope (n:=>:l) -> e l -> Maybe (e n)
```

Comment:
Good idea. If we just use the `Either (Name (n:=>:l))` naively then we'd just get the very first leaked name found. But maybe we should do `Either (Set (Name (n:=>:l)))`. Might be good for error reporting too. I'm hoping that being forced to use this will catch a lot of our leaked variable bugs and knowing their names will help fix them.